File: 프로그래머스/2/42587. 프로세스/프로세스.js
function solution(priorities, location) {
    let answer = 0;
    let queue = [...priorities]; // 깊은 복사
    let targetIndex = location; // 추적할 작업의 위치

    while (queue.length > 0) {
        // 큐의 첫 번째 작업 꺼내기
        let current = queue.shift();

        // 우선순위 높은 작업이 있다면 현재 작업을 뒤로 보냄
        if (queue.some(num => num > current)) {
            queue.push(current);
            // 추적 대상의 인덱스도 이동
            targetIndex = targetIndex === 0 ? queue.length - 1 : targetIndex - 1;
        } else {
            // 현재 작업 실행
            answer++;
            if (targetIndex === 0) {
                return answer; // 추적 대상 실행 순서 반환
            }
            targetIndex--;
        }
    }

    return answer;
}

File: 프로그래머스/2/42587. 프로세스/프로세스.cpp
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int solution(vector<int> priorities, int location) {
    queue<pair<int,int>> q;
    int n = priorities.size();

    for (int i = 0; i < n; i++) {
        q.push({priorities[i], i});
    }

    // 남은 문서들 중 "현재 최대 우선순위"를 쉽게 알기 위해 내림차순 정렬
    vector<int> sorted = priorities;
    sort(sorted.begin(), sorted.end(), greater<int>());

    int executed = 0; // 출력된 개수
    int k = 0;       // sorted[k] == 지금 출력되어야 하는 최대 우선순위

    while (!q.empty()) {
        auto cur = q.front();
        q.pop();

        // 지금 맨 앞 문서가 현재 최대 우선순위라면 출력
        if (cur.first == sorted[k]) {
            executed++;
            if (cur.second == location) return executed;
            k++;
        } 
        // 아니면 뒤로 보냄
        else {
            q.push(cur);
        }
    }

    return -1; // 논리상 도달하지 않음
}
